sdk/python/tests/compiler/compiler_tests.py

```diff
@@ -423,6 +423,38 @@ def my_pipeline(a: {'GCSPath': {'path_type':'file', 'file_type': 'csv'}}='good',
     finally:
       shutil.rmtree(tmpdir)
 
+  def test_type_checking_with_json_schema(self):
+    """Test type check pipeline parameters against the json schema."""
+    @component
+    def a_op(field_m: {'GCRPath': {'openapi_schema_validator': {"type": "string", "pattern": "^.*gcr\\.io/.*$"}}}, field_o: 'Integer'):
+      return ContainerOp(
+          name = 'operator a',
+          image = 'gcr.io/ml-pipeline/component-b',
+          arguments = [
+              '--field-l', field_m,
+              '--field-o', field_o,
+          ],
+      )
+
+    @pipeline(
+        name='p1',
+        description='description1'
+    )
+    def my_pipeline(a: {'GCRPath': {'openapi_schema_validator': {"type": "string", "pattern": "^.*gcr\\.io/.*$"}}}='good', b: 'Integer'=12):
+      a_op(field_m=a, field_o=b)
+
+    test_data_dir = os.path.join(os.path.dirname(__file__), 'testdata')
+    sys.path.append(test_data_dir)
+    tmpdir = tempfile.mkdtemp()
+    try:
+      simple_package_path = os.path.join(tmpdir, 'simple.tar.gz')
+      import jsonschema
+      with self.assertRaises(jsonschema.exceptions.ValidationError):
+        compiler.Compiler().compile(my_pipeline, simple_package_path, type_check=True)
```

Comment:
Maybe we can just call `_compile` which does not write anything to a file?

Reply:
I'm in fact in favor of using public APIs in the tests instead of private members. Testing the behaviors just as the users(public APIs) do can help us catch more potential bugs. For example, the bug might exist in the public API function that calls the private function. WDYT?